### DISCUSSION TITLE
The format and structure of the doc (vimdoc) are terrible and too different from modern projects' doc.

### DISCUSSION THREAD

**Discussion by sisrfeng** [vim/vim#13314] [Ideas]
- About "vimdoc's endless quirks": https://github.com/neovim/tree-sitter-vimdoc/issues/101
- The suggestion for neovim: https://github.com/neovim/neovim/issues/18017
(but "we deliberately take the documentation from Vim and only change it where Neovim diverges. There are no plans to rewrite the upstream documentation, as the increased cost of pulling in upstream changes would significantly outweigh the benefits of better readability" )
- My improved version of the doc: https://gitee.com/llwwff/doc_nvim 
It is clear for me but maybe not for others. So I just provide a rough reference. I hope that more people can pay attention to the work.

**Comment by Shane-XB-Qian** [comment]
no idea what are you talking about, but vim help doc was the greatest and clear/jump-able doc i ever see.
// vs though not so-called 'modern' (another so called 'modern' again?)? not sure who gave the standard about 'modern'

but i do see some modern doc was so mess, e.g mysql doc.

**Reply by sisrfeng** [reply]
I think the doc of vscode is modern and clear, though we can't jump to it using something like `:help XXX` .

**Reply by sisrfeng** [reply]
Do you think the 4 points described in `https://github.com/neovim/neovim/issues/18017#issue-1195820347` are pain points?

**Reply by chrisbra** [reply]
No, I don't think so.

1) Using a tabstop of 8 makes it perfectly readable on a terminal.
2) I don't see why this is an issue
3) table of content? we can add more, but I personally don't see this as an issue. I think most users will jump directly to the intersting topics using the helptags.
4) I don't think `vi_default` tags are annoying. I kind of like, that it's there so I even know what comes from the original vi and can be expected to work there as well. (and I guess this specifically only relates to the options.txt help file?

**Reply by sisrfeng** [reply]
1. Would tabstop of 4 be better? Now a sentence often can't fit into a line.
2. The help tags is usefull, but I think they should be hidden.

**Reply by sisrfeng** [reply]
I think a modern doc should looks like this, and that is what I have been try to convet the official vimdoc to, 
while keep the doc jump-able and readable in terminal.
![Uploading image.png…]()

**Reply by chrisbra** [reply]
1) I don't think so, because now you can use less to read the file perfectly fine in the terminal. I also don't see what's wrong with sentences taking more than one line
2) I don't think this is possible, not sure

**Reply by sisrfeng** [reply]
IMO, the ideal vimdoc should look like what is described in "Modified doc (with some annoying notation concealed)" in https://github.com/neovim/neovim/issues/18017#issuecomment-1192119037

**Reply by chrisbra** [reply]
I agree the vcode doc looks nice in a browser, since you can embed pictures and show the table of contents in a side bar. But that will be hard to do for a terminal application

**Reply by sisrfeng** [reply]
> I agree the vcode doc looks nice in a browser, since you can embed pictures and show the table of contents in a side bar. But that will be hard to do for a terminal application

I do not use pictures in my doc_nvim, and can have  table of contents in a side bar. Pls see the updated picture in my mentioned repo (https://gitee.com/llwwff/doc_nvim)

**Reply by ychin** [reply]
> Pls see the updated picture in my mentioned repo (https://gitee.com/llwwff/doc_nvim)

That example is very specific to **Neovim**. I'm not understanding what you are proposing regarding **Vim** documentation. Do you have an example? I don't disagree the documentation format could see some improvements, but your proposal are too ambiguous that it's hard to know exactly what you are suggesting.

---

1. 80-char limit is to make text readable. Columns that are too wide makes texts hard to read. This is why academic papers are written in multi-column format on a page. Tabstop 8 is an editor config, not specific to how documentation is written. Some of the docs are indented multiple levels deep, and some of them are indented using space instead of tabs. I think you should be more specific about exactly what you are asking for. Are you asking for reflowable text? A harder parsable syntax for writing docs? Use a larger column like 120-char instead of 80? Etc? Ambiguous suggestions is going to why you are getting a shrug from everyone.

2. I think the help tags are useful to show how that docs relate to other topics. I guess you could be super smart about it and selectively show/hide them, but they definitely need to be there in the source. I am still struggling to see how this is a big issue other than a few places where you see multiple lines of redundant help tags. Fixing these stuff takes time and this seems like a low-priority issue to me.

3. are you talking about browsing docs in TUI or the web? You are coming from Neovim, but their docs don't even have a table of contents in the terminal version of the docs (they do have them in the web version at https://neovim.io/doc/)

4. Default values: that's not really Vim's problem (since you are discussing Neovim). The default values are *really* useful to know for the user. You could argue that a Windows user doesn't care about default GTK values, but most options have a single default anyway. You *could* add a lot of generators to build platform-specific documentation pages but that seems like a lot of work for little gains. If Neovim's options have discrepancies from Vim in default values, they should fix their docs.

(Edit: Ok, re-reading your point you were talking about the Vi defaults, not Vim. I don't know it just seems like such a minor issue to me 🤷)

---

I'm not saying Vim docs are perfect, but I think we need to take into account for cost/benefits analysis, and also I think we would appreciate more concrete suggestions. Some of what you mentioned like tabstop=8 are editor configuration but you seem to be complaining about the formatting instead, and some of the complaints you have seem Neovim-specific.

**Comment by chrisbra** [comment]
you improved version throws 403. It's not clear what problems you are facing. If you want to come up with concreate improvements, please send a PR.

**Reply by sisrfeng** [reply]
Sorry for 403. Now https://gitee.com/llwwff/doc_nvim is accessible. I'll try to send a PR later.

**Comment by Shane-XB-Qian** [comment]
if you meant doc in webpage is 'modern', there should be some site offer that already, you can gg(? :lol:) it.

--
shane.xb.qian

**Reply by sisrfeng** [reply]
Pls see https://github.com/vim/vim/discussions/13314#discussioncomment-7250429

**Comment by Shane-XB-Qian** [comment]
> I agree the vcode doc looks nice in a browser, since you can embed pictures and show the table of contents in a side bar. But that will be hard to do for a terminal application

perhaps it's not about 'hard' (some 3rd part website had/could do that), vs/but according to develop guidleline seems VIM should not going to be a full GUI tool or something, that said it's not vim's goal.

--
shane.xb.qian

**Reply by sisrfeng** [reply]
Again, pls see https://github.com/vim/vim/discussions/13314#discussioncomment-7250429

**Comment by sisrfeng** [comment]
> what's wrong with sentences taking more than one line

pls see https://sembr.org/
```
And what 
do 
you feel
whlie you are
reading some
lines like
this ?
I think such
lines make it 
hard to
understand what
they say!
```

I haven't seen so much short and "broken" lines in docs of other projects. Even the old school `man` page can display longer lines.

As for nvim, (see https://github.com/neovim/neovim/issues/18017#issuecomment-1757684839) 
> some have "flow" layout (instead of fixed-width lines), for example https://neovim.io/doc/user/lua.html

Linus: (see https://lkml.org/lkml/2020/5/29/1038) 
> I don't want to see patches that make the kernel reading
experience worse for me and likely for the vast majority of people,
based on the argument that some odd people have small terminal
windows.

**Reply by chrisbra** [reply]
so when do you have a help window with only 20 character width?

**Reply by sisrfeng** [reply]
This is just an extreme example to show why small character width can be a problem.

**Reply by sisrfeng** [reply]
As https://sembr.org/ puts it 
> Why Use Semantic Line Breaks?
> For a writer, semantic line breaks allow the physical structure of text to reflect the logical structure of the thoughts that produce them.

**Reply by sisrfeng** [reply]
A example :
![image](https://github.com/vim/vim/assets/53520949/1ee7e225-ef94-4005-9f95-06b46bee1438)
The word `sorted` is at another line. It's so ugly and breaks the physical structure of text.

**Reply by chrisbra** [reply]
true, it should fit into the previous line and then we should add an empty line after that. That's one concrete example, for which we should create a PR :)

**Reply by chrisbra** [reply]
It's seems to be caused by concealing. If you disable it, you'll see that `sorted.` breaks the max line with and therefore was added into a new line.

**Reply by sisrfeng** [reply]
Oh I see. This is also a concrete example showing why an indentation of 8 space and the 79 character limitation may cause problems.

**Reply by ychin** [reply]
> some have "flow" layout (instead of fixed-width lines), for example https://neovim.io/doc/user/lua.html

This statement needs some unpacking for complete context for those who don't know how it works.

The documentation page for lua is still fixed-width when viewed within Neovim ([lua.txt](https://github.com/neovim/neovim/blob/master/runtime/doc/lua.txt)). What Neovim did was that they wrote a tree-sitter docs parser (https://github.com/neovim/tree-sitter-vimdoc/) that could parse vim docs, and use it to output a reflowable online web documentation with sans-serif texts which are more readable than monospace and adopt to arbitrary column width. I think it's a good idea. To do that though, they had to either write new documentation in such pre-defined formats, or fix up old docs, because the tree-sitter parser is quite rigid and you have to follow explicit rules in order for it to work. That's why they only turned on "new style" / reflowable mode for certain docs known to be written with this syntax in mind (e.g. lua.txt, api.txt).

I actually did some investigation into using it for MacVim docs before (see [link](https://github.com/neovim/neovim/pull/20690#issuecomment-1458796967) in case people are interested) but decided against it, because you kind of have to really buy in to the syntax, and make certain compromises (e.g. it doesn't support tables, so if you make a table you have to wrap things in a code block which is not ideal, see https://neovim.io/doc/user/api.html#api-definitions).

But the general idea is something I liked. Basically, keep the source vim doc in a pre-formatted 80-char, but make sure the syntax is written in a way that can be used to generate a reflowable web version. The only other way to do this is to write the documentation in say Markdown/LaTeX/HTML/reStructuredText/Typst as source and then generate vim doc / HTML from that. It's definitely a lot of work though.

> It's seems to be caused by concealing. If you disable it, you'll see that sorted. breaks the max line with and therefore was added into a new line.

Kind of related old issue (it has more to do with 'wrap' behaviors with concealed texts) that was closed :/ https://github.com/vim/vim/issues/260

**Reply by ychin** [reply]
>  The word sorted is at another line. It's so ugly and breaks the physical structure of text.

This will happen no matter what width you use. You are going to have a straggler that is the lone word at the next line. We don't have infinite width monitors.

---

FWIW the example above has the texts description for say "helptags" indented 3 levels deep (3 x 8 = 24 chars out of an 80-char column). I think we could reformat them to use fewer indentation levels so they don't end up being just 79-24=55-char columns.

**Comment by sisrfeng** [comment]
> keep the source vim doc in a pre-formatted 80-char
I know it cost lots of time to rewrite all the docs, but is it necessary to keep the 80-char witdh in further new content ?

**Comment by sisrfeng** [comment]
> This will happen no matter what width you use. You are going to have a straggler that is the lone word at the next line. We don't have infinite width monitors.

It is just an example which is against the https://sembr.org/. Maybe this example (as I mentioned above) is more opvious:
```
And what 
do 
you feel
whlie you are
reading some
lines like
this ?
I think such
lines make it 
hard to
understand what
they say!
```
